daml/Marketplace/Trading/Model.daml

```diff
@@ -63,6 +63,12 @@ data Execution = Execution with
 
 type T = Order
 
+data TradeCollateral = Cleared { clearinghouse: Party } | Collateral (ContractId AssetDeposit) deriving (Show, Eq)
```

Comment:
```suggestion
data TradeCollateral 
  = Cleared with clearinghouse: Party 
  | Collateral (ContractId AssetDeposit) 
    deriving (Show, Eq)
```

Reply:
Are these really mutually exclusive? Couldn't you have a cleared trade that requires collateral also?

Reply:
Hmmm not 100% sure. I figured with a cleared trade in this context, the "collateral" is the membership in the clearinghouse and whatever accounts you hold with them. Maybe `Collateral` vs `Noncollateralized` might make sense? 

In that case, instead of passing the clearinghouse around with the order, the order choice could pull up the listing and check if a clearing membership/approval is necessary instead, regardless of if a collateral is used or not.

Reply:
Actually, looking up the listing in the order creation choice is going to require multi-party submits in daml hub since it will require `readAs: [public]`, so I don't think this will be possible yet.

Reply:
@brian-weir @lucianojoublanc-da - I think a good solution to this, as well as for better separation of the clearing/collateral concerns can be to rely on the `Listing` for making decisions such as: what collateral is necessary, if a clearinghouse needs to approve the trade, how the trade gets settled, etc. Which makes sense to me, since the listing is where all of that should be specified anyway.

So in this case: `TradeCollateral` could just be something like `optCollateral: Optional (ContractId AssetDeposit)` (or still a special type for readability), and the order request choice can lookup the listing information and see if a collateral is required and reject the order if it does not exist. If the listing requires the order to be cleared by a clearinghouse, that can happen there as well. This would also allow for an order to require collateral _and_ require clearing. Then once the orders are matched and get to the matching service, the matching service can get the listing information as well and process the trade accordingly.

If that seems like a good solution, then the question for implementing it is: should the `Order` have the `ContractId Listing`, the `Listing`, or just the relevant info? If it's the contract id or key info, that won't currently work on hub since it would require `readAs: [publicParty]` since the `customer` isn't a direct observer on a `Listing`. Let me know what you think, I'm open to other solutions as well.

Reply:
I do like the idea of linking this with the listing somehow. It makes sense from a business point of view, as I think it's where this would be determined in the real world. So you could originate the same instrument and then have it traded multiple ways depending on the listing.

However this will prevent e.g. a market maker from clearing the trade vs. a user trading fully collateralised on the same order-book; you'll have two separate instruments. I think this scenario was an original business requirement.

> that won't currently work on hub since it would require readAs: [publicParty] since the customer isn't a direct observer on a Listing

So how does it currently have the asset details in there? I assume the `Asset` is just embeded in `Order`?
Yeah, that does seem like a problem.

This sounds like a considerable refactor. Is it worth doing this separately, so we don't hold up this merge? I do definitely think that it's worth doing it though.

Reply:
@alexgraham-da - I think your idea is sound. Ideally, the `Trading` service shouldn't care if the order is traded on margin, cleared, fully collateralised, etc. It's a much cleaner separation if we move these details onto the `Listing` contract and have these `Pre-Trade` checks against an order request live on the `Listing` service. There are some existing TODOs around checking precision, max/min quantity and even the status of the listing against an order.

> So how does it currently have the asset details in there? I assume the Asset is just embeded in Order?

Currently, an order is linked to a listing via the listings 'symbol' (which is also part of the contract key on `Listing` as `listingId` which allows for lookups). Sending the same symbol twice to Exberry will cause an error. 

> However this will prevent e.g. a market maker from clearing the trade vs. a user trading fully collateralised on the same order-book; you'll have two separate instruments. I think this scenario was an original business requirement.

To satisfy this requirement, we'll need a way of identifying market-makers from other users who must put up full collateral for a trade. There's multiple ways of implementing this requirement but I don't think we should have multiple listings/instruments. I think it'll require a bit more design and should be part of another PR. I've added this point to the spreadsheet so we have it tracked there for now.

Reply:
> Currently, an order is linked to a listing via the listings 'symbol' (which is also part of the contract key on Listing as listingId which allows for lookups).

I was looking at this, and I don't think this will work 100% for looking up a `Listing` in all places - the key of the `Listing` is `(operator, provider, listingId)`. Since the provider of the `Trading` service isn't necessarily the provider of the `Listing`, you wouldn't have a way of knowing which `provider` to use for the key. Maybe it would make more sense for a `Listing` to be keyed by an `Id`? That way you can store the `Id` in contracts like `Order` that need to do a lookup.

Reply:
> Ideally, the Trading service shouldn't care if the order is traded on margin, cleared, fully collateralised, etc. It's a much cleaner separation if we move these details onto the Listing contract and have these Pre-Trade checks against an order request live on the Listing service. There are some existing TODOs around checking precision, max/min quantity and even the status of the listing against an order.

I think then this is indeed the way to go. It's not a _huge_ refactor to do just this, but I think it makes the most sense to get this PR out now and then refactor and add this functionality separately (especially since we'll need to have a way of handling visibility issues with the `Listing` on Hub).

The market maker vs. regular user requirement is an interesting one - I'd say that's definitely separate and we can decide later if it would need to be rolled into the listing pre-trade functionality at the same time, or just design that in a way that would eventually allow it.